README.md
hdkey
=====

A Dart library for [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki) (hierarchical deterministic keys).

Installation
------------

    dart pub add hdkey

Usage
-----

**example:**

```dart
import 'dart:typed_data';
import 'package:hdkey/hdkey.dart';
import 'package:hex/hex.dart';

void main() {
  const seed = 'a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af';
  final HDKey hdkey = HDKey.fromMasterSeed(Uint8List.fromList(HEX.decode(seed)));
  print(hdkey.privateExtendedKey);
  // output => 'xprv9s21ZrQH143K2SKJK9EYRW3Vsg8tWVHRS54hAJasj1eGsQXeWDHLeuu5hpLHRbeKedDJM4Wj9wHHMmuhPF8dQ3bzyup6R7qmMQ1i1FtzNEW'
  print(hdkey.publicExtendedKey);
  // output => 'xpub661MyMwAqRbcEvPmRAmYndzERhyNux1GoHzHxgzVHMBFkCro3kbbCiDZZ5XabZDyXPj5mH3hktvkjhhUdCQxie5e1g4t2GuAWNbPmsSfDp2'
}
```

### `HDKey.fromMnemonic(String mnemonic, [Bip32Type? this._versions])`

Creates an `HDKey` object from a mnemonic string. Accepts an optional `versions` object.

```dart
const mnemonic = 'chapter eager old retreat wire drift deal later ignore magic veteran liberty crime rice describe';
final HDKey hdkey = HDKey.fromMasterSeed(mnemonic);
```

### `HDKey.fromMasterSeed(Uint8List seed, [Bip32Type? this._versions])`

Creates an `HDKey` object from a master seed. Accepts an optional `versions` object.

```dart
const seed = 'a0c42a9c3ac6abf2ba6a9946ae83af18f51bf1c9fa7dacc4c92513cc4dd015834341c775dcd4c0fac73547c5662d81a9e9361a0aac604a73a321bd9103bce8af';
final HDKey hdkey = HDKey.fromMasterSeed(Uint8List.fromList(HEX.decode(seed)));
```

### `HDKey.fromExtendedKey(String base58key, [Bip32Type? this._versions])`

Creates an `HDKey` object from a `xprv` or `xpub` extended key string. Accepts an optional `versions` object.

```dart
const key = 'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j';
final HDKey hdkey = HDKey.fromExtendedKey(key);
```

**or**

```dart
const key = 'xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt';
final HDKey hdkey = HDKey.fromExtendedKey(key);
```

### `HDKey derive(String path)`

Derives the `HDKey` at `path` from the current `HDKey`.

```dart
var seed =
    'fffcf9f6f3f0edeae7e4e1dedbd8d5d2cfccc9c6c3c0bdbab7b4b1aeaba8a5a29f9c999693908d8a8784817e7b7875726f6c696663605d5a5754514e4b484542';
final HDKey hdkey = HDKey.fromMasterSeed(Uint8List.fromList(HEX.decode(seed)));

HDKey childkey = hdkey.derive("m/0/2147483647'/1");

print(childkey.privateExtendedKey);
// -> "xprv9zFnWC6h2cLgpmSA46vutJzBcfJ8yaJGg8cX1e5StJh45BBciYTRXSd25UEPVuesF9yog62tGAQtHjXajPPdbRCHuWS6T8XA2ECKADdw4Ef"
print(childkey.publicExtendedKey);
// -> "xpub6DF8uhdarytz3FWdA8TvFSvvAh8dP3283MY7p2V4SeE2wyWmG5mg5EwVvmdMVCQcoNJxGoWaU9DCWh89LojfZ537wTfunKau47EL2dhHKon"
```
### `Uint8List sign(Uint8List hash)`

Signs the Uint8List `hash` with the private key using `secp256k1` and returns the signature as a Uint8List.

### `bool verify(Uint8List hash, Uint8List signature)`

Verifies that the `signature` is valid for `hash` and the `HDKey`'s public key using `secp256k1`. Returns `true` for
valid, `false` for invalid. Throws if the `hash` or `signature` is the wrong length.

### `HDKey wipePrivateData()`

Wipes all record of the private key from the `HDKey` instance. After calling this method, the instance will behave as if
it was created via `HDKey.fromExtendedKey(xpub)`.

### `Uint8List? get privateKey`

Getter/Setter of the `HDKey`'s private key, stored as a Uint8List.

### `Uint8List? get publicKey`

Getter/Setter of the `HDKey`'s public key, stored as a Uint8List.

### `String? get privateExtendedKey`

Getter/Setter of the `HDKey`'s `xprv`, stored as a string.

### `String? get publicExtendedKey`

Getter/Setter of the `HDKey`'s `xpub`, stored as a string.

### `ExtendedKeys get extendedKeys`

Getter/Setter of the `HDKey`'s `xpub` and `xprv`, stored as a `ExtendedKeys`.

References
----------
This library is the development of the [hdkey](https://github.com/cryptocoinjs/hdkey) library developed for Javascript for Dart.

License
-------

MIT

CHANGELOG.md
## 0.0.1

- Initial version, created by Mahsum UREBE

test/hdkey_test.dart
import 'dart:typed_data';

import 'package:bs58check/bs58check.dart' as bs58check;
import 'package:hdkey/hdkey.dart';
import 'package:hdkey/src/types.dart';
import 'package:hex/hex.dart';
import 'package:test/test.dart';

import 'fixtures/hdkey.dart';

void main() {
  group('HDKey', () {
    group('+ fromMasterSeed', () {
      fixtures[FixtureKeys.valid]?.forEach((f) {
        test('should properly derive the chain path: $f.path', () {
          HDKey hdKey = HDKey.fromMasterSeed(f.seed);
          HDKey child = hdKey.derive(f.path);

          expect(
              bs58check.decode(child.privateExtendedKey!), equals(f.private));
          expect(bs58check.decode(child.publicExtendedKey), equals(f.public));
        });

        group('> ' + f.path + ' extendedKey / fromExtendedKey', () {
          test('should return an ExtendedKeys read for HDKey serialization',
              () {
            HDKey hdKey = HDKey.fromMasterSeed(f.seed);
            HDKey child = hdKey.derive(f.path);
            ExtendedKeys extendedKeys = ExtendedKeys(
                bs58check.encode(f.public), bs58check.encode(f.private));
            expect(child.extendedKeys, extendedKeys);

            HDKey newKey = HDKey.fromExtendedKey(extendedKeys.private!);
            expect(newKey.privateExtendedKey, bs58check.encode(f.private));
            expect(newKey.publicExtendedKey, bs58check.encode(f.public));
          });
        });
      });
    });
    group('- privateKey', () {
      test('should throw an error if is null', () {
        HDKey hdkey = new HDKey();
        expect(() => hdkey.privateKey = null, throwsFormatException);
      });
      test('should throw an error if incorrect key size', () {
        HDKey hdkey = new HDKey();
        expect(() => hdkey.privateKey = Uint8List.fromList([1, 2, 3, 4]),
            throwsFormatException);
      });
    });
    group('- publicKey', () {
      test('should throw an error if is null', () {
        HDKey hdkey = new HDKey();
        expect(() => hdkey.publicKey = null, throwsFormatException);
      });
      test('should throw an error if incorrect key size', () {
        HDKey hdkey = new HDKey();
        expect(() => hdkey.publicKey = Uint8List.fromList([1, 2, 3, 4]),
            throwsFormatException);
      });
      test('should not throw if key is 33 bytes (compressed)', () {
        HDKey hdkey = new HDKey();

        final publicKey = Uint8List.fromList(HEX.decode(
            '039b50417d0a067207194dd4239fdfcd70bdcd609e46880df52e64107ec0fed4b7'));
        try {
          hdkey.publicKey = publicKey;
          expect(hdkey.publicKey, equals(publicKey));
        } catch (e) {
          fail(e.toString());
        }
      });
      test('should not throw if key is 65 bytes (not compressed)', () {
        HDKey hdkey = new HDKey();

        final publicKeyNotCompressed = Uint8List.fromList(HEX.decode(
            '04a3b011a2ef2994e3957009313cce00f8463e843b83fb95996a63d6c0bae728865579e878d43e06807cfca494c0b618ef5cd37f3a2bd5611dc3037e44f28de166'));
        final publicKey = Uint8List.fromList(HEX.decode(
            '02a3b011a2ef2994e3957009313cce00f8463e843b83fb95996a63d6c0bae72886'));
        try {
          hdkey.publicKey = publicKeyNotCompressed;
          expect(hdkey.publicKey, equals(publicKey));
        } catch (e) {
          fail(e.toString());
        }
      });
    });

    group('+ fromExtendedKey()', () {
      group('> when private', () {
        test('should parse it', () {
          // m/0/2147483647'/1/2147483646'/2;
          const String key =
              'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j';
          HDKey hdkey = HDKey.fromExtendedKey(key);
          expect(hdkey.versions.private, equals(0x0488ade4));
          expect(hdkey.versions.public, equals(0x0488b21e));
          expect(hdkey.depth, equals(5));
          expect(hdkey.parentFingerprint, equals(0x31a507b8));
          expect(hdkey.index, equals(2));
          expect(
              HEX.encode(hdkey.chainCode!.toList()),
              equals(
                  '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'));
          expect(
              HEX.encode(hdkey.privateKey!.toList()),
              equals(
                  'bb7d39bdb83ecf58f2fd82b6d918341cbef428661ef01ab97c28a4842125ac23'));
          expect(
              HEX.encode(hdkey.publicKey!.toList()),
              equals(
                  '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'));
          expect(HEX.encode(hdkey.identifier!.toList()),
              equals('26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'));
        });
      });
      group('> when public', () {
        test('should parse it', () {
          // m/0/2147483647'/1/2147483646'/2;
          const String key =
              'xpub6FnCn6nSzZAw5Tw7cgR9bi15UV96gLZhjDstkXXxvCLsUXBGXPdSnLFbdpq8p9HmGsApME5hQTZ3emM2rnY5agb9rXpVGyy3bdW6EEgAtqt';
          HDKey hdkey = HDKey.fromExtendedKey(key);
          expect(hdkey.versions.private, equals(0x0488ade4));
          expect(hdkey.versions.public, equals(0x0488b21e));
          expect(hdkey.depth, equals(5));
          expect(hdkey.parentFingerprint, equals(0x31a507b8));
          expect(hdkey.index, equals(2));
          expect(
              HEX.encode(hdkey.chainCode!.toList()),
              equals(
                  '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'));
          expect(hdkey.privateKey, isNull);
          expect(
              HEX.encode(hdkey.publicKey!.toList()),
              equals(
                  '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'));
          expect(HEX.encode(hdkey.identifier!.toList()),
              equals('26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'));
        });
      });
    });
    final validFixtures = fixtures[FixtureKeys.valid]!;
    group('> when signing', () {
      test('should work', () {
        const String key =
            'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j';
        String aHex =
            '6ba4e554457ce5c1f1d7dbd10459465e39219eb9084ee23270688cbe0d49b52b7905d5beb28492be439a3250e9359e0390f844321b65f1a88ce07960dd85da06';
        String bHex =
            'dfae85d39b73c9d143403ce472f7c4c8a5032c13d9546030044050e7d39355e47a532e5c0ae2a25392d97f5e55ab1288ef1e08d5c034bad3b0956fbbab73b381';

        HDKey hdkey = HDKey.fromExtendedKey(key);
        Uint8List ma = Uint8List.fromList(List.filled(32, 0));
        Uint8List mb = Uint8List.fromList(List.filled(32, 8));
        Uint8List a = hdkey.sign(ma);
        Uint8List b = hdkey.sign(mb);
        expect(HEX.encode(a), equals(aHex));
        expect(HEX.encode(b), equals(bHex));
        expect(hdkey.verify(ma, a), equals(true));
        expect(hdkey.verify(mb, b), equals(true));
        expect(hdkey.verify(Uint8List(32), Uint8List(64)), equals(false));
        expect(hdkey.verify(ma, b), equals(false));
        expect(hdkey.verify(mb, a), equals(false));

        expect(() => hdkey.verify(Uint8List(99), a), throwsArgumentError);
        expect(() => hdkey.verify(ma, Uint8List(99)), throwsArgumentError);
      });
      group('> when deriving public key', () {
        test('should work', () {
          const String key =
              'xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8';
          HDKey hdkey = HDKey.fromExtendedKey(key);
          const String path = 'm/3353535/2223/0/99424/4/33';
          HDKey derivedHDKey = hdkey.derive(path);
          const String expected =
              'xpub6JdKdVJtdx6sC3nh87pDvnGhotXuU5Kz6Qy7Piy84vUAwWSYShsUGULE8u6gCivTHgz7cCKJHiXaaMeieB4YnoFVAsNgHHKXJ2mN6jCMbH1';
          expect(derivedHDKey.publicExtendedKey, equals(expected));
        });
      });
      group('> when private key integer is less than 32 bytes', () {
        test('should work', () {
          const String seed = '000102030405060708090a0b0c0d0e0f';
          HDKey masterKey =
              HDKey.fromMasterSeed(Uint8List.fromList(HEX.decode(seed)));
          HDKey newKey = masterKey.derive("m/44'/6'/4'");
          const String expected =
              'xprv9ymoag6W7cR6KBcJzhCM6qqTrb3rRVVwXKzwNqp1tDWcwierEv3BA9if3ARHMhMPh9u2jNoutcgpUBLMfq3kADDo7LzfoCnhhXMRGX3PXDx';
          expect(newKey.privateExtendedKey, equals(expected));
        });
      });
      group('HARDENED_OFFSET', () {
        test('should be set', () {
          expect(HDKey.HARDENED_OFFSET, equals(0x80000000));
        });
      });
      group('> when private key has leading zeros', () {
        test('will include leading zeros when hashing to derive child', () {
          const String key =
              'xprv9s21ZrQH143K3ckY9DgU79uMTJkQRLdbCCVDh81SnxTgPzLLGax6uHeBULTtaEtcAvKjXfT7ZWtHzKjTpujMkUd9dDb8msDeAfnJxrgAYhr';
          HDKey hdkey = HDKey.fromExtendedKey(key);
          expect(
              HEX.encode(hdkey.privateKey!),
              equals(
                  '00000055378cf5fafb56c711c674143f9b0ee82ab0ba2924f19b64f5ae7cdbfd'));
          HDKey derived = hdkey.derive("m/44'/0'/0'/0/0'");
          expect(
              HEX.encode(derived.privateKey!),
              equals(
                  '3348069561d2a0fb925e74bf198762acc47dce7db27372257d2d959a9e6f8aeb'));
        });
      });
      group('> when private key is null', () {
        test('privateExtendedKey should return null and not throw', () {
          const String seed = '000102030405060708090a0b0c0d0e0f';
          HDKey masterKey =
              HDKey.fromMasterSeed(Uint8List.fromList(HEX.decode(seed)));
          expect(masterKey.privateExtendedKey, isNotNull);
          masterKey.wipePrivateData();
          try {
            masterKey.privateExtendedKey;
          } catch (e) {
            fail(e.toString());
          }
          expect(masterKey.privateExtendedKey, isNull);
        });
      });
      group(
          ' - when the path given to derive contains only the master extended key',
          () {
        HDKey hdKeyInstance = HDKey.fromMasterSeed(validFixtures[0].seed);
        test('should return the same hdkey instance', () {
          expect(hdKeyInstance.derive('m'), equals(hdKeyInstance));
          expect(hdKeyInstance.derive('M'), equals(hdKeyInstance));
          expect(hdKeyInstance.derive("m'"), equals(hdKeyInstance));
          expect(hdKeyInstance.derive("M'"), equals(hdKeyInstance));
        });
      });
      group(
          ' - when the path given to derive does not begin with master extended key',
          () {
        test('should throw an error', () {
          expect(() => HDKey().derive('123'), throwsFormatException);
        });
      });
      group('- after wipePrivateData()', () {
        test('should not have private data', () {
          HDKey hdkey =
              HDKey.fromMasterSeed(validFixtures[6].seed).wipePrivateData();
          expect(hdkey.privateKey, isNull);
          expect(hdkey.privateExtendedKey, isNull);
          expect(() => hdkey.sign(Uint8List(32)), throwsException);
          final HDKey childKey = hdkey.derive('m/0');
          expect(childKey.publicExtendedKey, equals(bs58check.encode(validFixtures[7].public)));
          expect(childKey.privateKey, isNull);
          expect(childKey.privateExtendedKey, isNull);
        });
        test('should have correct data', () {
          // m/0/2147483647'/1/2147483646'/2;
          const String key =
              'xprvA2nrNbFZABcdryreWet9Ea4LvTJcGsqrMzxHx98MMrotbir7yrKCEXw7nadnHM8Dq38EGfSh6dqA9QWTyefMLEcBYJUuekgW4BYPJcr9E7j';
          HDKey hdkey = HDKey.fromExtendedKey(key).wipePrivateData();
          expect(hdkey.versions.private, equals(0x0488ade4));
          expect(hdkey.versions.public, equals(0x0488b21e));
          expect(hdkey.depth, equals(5));
          expect(hdkey.parentFingerprint, equals(0x31a507b8));
          expect(hdkey.index, equals(2));
          expect(
              HEX.encode(hdkey.chainCode!),
              equals(
                  '9452b549be8cea3ecb7a84bec10dcfd94afe4d129ebfd3b3cb58eedf394ed271'));
          expect(
              HEX.encode(hdkey.publicKey!),
              equals(
                  '024d902e1a2fc7a8755ab5b694c575fce742c48d9ff192e63df5193e4c7afe1f9c'));
          expect(HEX.encode(hdkey.identifier!),
              equals('26132fdbe7bf89cbc64cf8dafa3f9f88b8666220'));
        });
        test('should be able to verify signatures', () {
          final HDKey fullKey = HDKey.fromMasterSeed(validFixtures[0].seed);
          // using JSON methods to clone before mutating;
          final HDKey wipedKey =
              HDKey.fromExtendedKey(fullKey.privateExtendedKey!)
                  .wipePrivateData();
          final Uint8List hash = Uint8List.fromList(List.filled(32, 8));
          expect(wipedKey.verify(hash, fullKey.sign(hash)), isTrue);
        });
        test('should not throw if called on hdkey without private data', () {
          HDKey hdkey =
              HDKey.fromExtendedKey(bs58check.encode(validFixtures[0].public));
          try {
            hdkey.wipePrivateData();
          } catch (e) {
            fail(e.toString());
          }
          expect(hdkey.publicExtendedKey,
              bs58check.encode(validFixtures[0].public));
        });
      });
      group('Deriving a child key does not mutate the internal state', () {
        test('should not mutate it when deriving with a private key', () {
          HDKey hdkey =
              HDKey.fromExtendedKey(bs58check.encode(validFixtures[0].private));
          const String path = 'm/123';
          final String privateKeyBefore = HEX.encode(hdkey.privateKey!);
          HDKey child = hdkey.derive(path);
          expect(HEX.encode(hdkey.privateKey!), equals(privateKeyBefore));
          HDKey child2 = hdkey.derive(path);
          expect(HEX.encode(hdkey.privateKey!), equals(privateKeyBefore));
          HDKey child3 = hdkey.derive(path);
          expect(HEX.encode(hdkey.privateKey!), equals(privateKeyBefore));
          expect(child.privateKey, equals(child2.privateKey));
          expect(child2.privateKey, equals(child3.privateKey));
        });
        test('should not mutate it when deriving without a private key', () {
          HDKey hdkey =
              HDKey.fromExtendedKey(bs58check.encode(validFixtures[0].private));
          const String path = 'm/123/123/123';
          hdkey.wipePrivateData();
          String publicKeyBefore = HEX.encode(hdkey.publicKey!);
          HDKey child = hdkey.derive(path);
          expect(HEX.encode(hdkey.publicKey!), equals(publicKeyBefore));
          HDKey child2 = hdkey.derive(path);
          expect(HEX.encode(hdkey.publicKey!), equals(publicKeyBefore));
          HDKey child3 = hdkey.derive(path);
          expect(HEX.encode(hdkey.publicKey!), equals(publicKeyBefore));
          expect(HEX.encode(child.publicKey!),
              equals(HEX.encode(child2.publicKey!)));
          expect(HEX.encode(child2.publicKey!),
              equals(HEX.encode(child3.publicKey!)));
        });
      });
    });
  });
}


test/helpers/common_helper_test.dart
import 'package:hdkey/src/helpers/common_helper.dart';
import 'package:test/test.dart';

void main() {
  group('Common Functions', () {
    group('trueOrThrow', () {
      test('Should be thrown', () {
        expect(() => trueOrThrow(false, Exception()), throwsException);
      });
      test('Should not be thrown', () {
        try {
          trueOrThrow(true, Exception());
          expect(true, isTrue);
        } catch (e) {
          fail('Should not be thrown');
        }
      });
    });
  });
}


lib/src/hdkey.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:bip32/bip32.dart';
import 'package:bip39/bip39.dart' as bip39;
import 'package:bs58check/bs58check.dart';
import 'package:bs58check/bs58check.dart' as bs58check;
import 'package:hex/hex.dart';

import './helpers/common_helper.dart';
import './helpers/crypto_helper.dart';
import './types.dart';

final BITCOIN_VERSION = new Bip32Type(public: 0x0488B21E, private: 0x0488ADE4);
final MASTER_SECRET = Uint8List.fromList(utf8.encode('Bitcoin seed'));
const LEN = 78;

class HDKey {
  static const HARDENED_OFFSET = 0x80000000;

  Bip32Type get versions {
    return this._versions ?? BITCOIN_VERSION;
  }

  Bip32Type? _versions;
  int depth = 0;
  int index = 0;

  Uint8List? _privateKey;
  Uint8List? _publicKey;
  Uint8List? chainCode;
  int _fingerprint = 0;
  int parentFingerprint = 0;
  Uint8List? _identifier;

  //region Getters
  int get fingerprint => _fingerprint;

  Uint8List? get identifier => _identifier;

  Uint8List? get pubKeyHash => _identifier;

  Uint8List? get privateKey => _privateKey;

  Uint8List? get publicKey => _publicKey;

  String? get privateExtendedKey {
    if (this._privateKey != null) {
      List<int> key = [];
      key.add(0);
      key.addAll(this.privateKey!.toList());
      return bs58check
          .encode(_serialize(this.versions.private, Uint8List.fromList(key)));
    }
    return null;
  }

  String get publicExtendedKey {
    return bs58check.encode(_serialize(this.versions.public, this.publicKey!));
  }

  ExtendedKeys get extendedKeys {
    return ExtendedKeys(this.publicExtendedKey, this.privateExtendedKey);
  }

  //endregion

  //region Setters

  void set privateKey(Uint8List? value) {
    trueOrThrow(
        value != null, FormatException('Private key must be defined.', value));
    trueOrThrow(value?.length == 32,
        FormatException('Private key must be 32 bytes.', value));
    trueOrThrow(privateKeyVerify(value!) == true,
        FormatException('Invalid private key', value));

    this._privateKey = value;
    this._publicKey = publicKeyCreate(value, true);
    this._identifier = hash160(this.publicKey!);
    this._fingerprint =
        this._identifier?.sublist(0, 4).buffer.asByteData().getUint32(0) ?? 0;
  }

  void set publicKey(Uint8List? value) {
    trueOrThrow(
        value != null, FormatException('Public key must be defined.', value));
    trueOrThrow(value?.length == 33 || value?.length == 65,
        FormatException('Public key must be 33 or 65 bytes.', value));
    trueOrThrow(publicKeyVerify(value!) == true,
        FormatException('Invalid public key', value));

    this._publicKey = publicKeyConvert(value, true); // force compressed point
    this._identifier = hash160(this.publicKey!);
    this._fingerprint =
        this._identifier!.sublist(0, 4).buffer.asByteData().getUint32(0);
    this._privateKey = null;
  }

  //endregion

  //region Constructors
  HDKey([Bip32Type? this._versions]);

  HDKey.fromMnemonic(String mnemonic, [Bip32Type? this._versions]) {
    Uint8List seedBuffer = bip39.mnemonicToSeed(mnemonic);
    final I = sha512(seedBuffer, MASTER_SECRET);
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);
    this.chainCode = IR;
    this.privateKey = IL;
  }

  HDKey.fromMasterSeed(Uint8List seed, [Bip32Type? this._versions]) {
    final I = sha512(seed, MASTER_SECRET);
    HEX.encode(I);
    final IL = I.sublist(0, 32);
    final IR = I.sublist(32);

    this.chainCode = IR;
    this.privateKey = IL;
  }

  HDKey.fromExtendedKey(String base58key, [Bip32Type? this._versions]) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    final decodedKeyBuffer = bs58check.decode(base58key);
    final keyBuffer = ByteData.view(decodedKeyBuffer.buffer);
    final version = keyBuffer.getUint32(0);

    trueOrThrow(
        version == versions.private || version == versions.public,
        FormatException(
            'Version mismatch: does not match private or public', base58key));

    this.depth = keyBuffer.getUint8(4);
    this.parentFingerprint = keyBuffer.getUint32(5);
    this.index = keyBuffer.getUint32(9);
    this.chainCode = decodedKeyBuffer.sublist(13, 45);

    var key = ByteData.view(decodedKeyBuffer.sublist(45).buffer);
    if (key.getUint8(0) == 0) {
      // private
      trueOrThrow(
          version == versions.private,
          FormatException(
              'Version mismatch: version does not match private', base58key));
      this.privateKey =
          Uint8List.view(key.buffer).sublist(1); // cut off first 0x0 byte
    } else {
      trueOrThrow(
          version == versions.public,
          FormatException(
              'Version mismatch: version does not match public', base58key));
      this.publicKey = Uint8List.view(key.buffer);
    }
  }

  //endregion

  //region Deriving Methods
  HDKey derive(String path) {
    if (path == 'm' || path == 'M' || path == "m'" || path == "M'") {
      return this;
    }

    List<String> entries = path.split('/');
    HDKey hdKey = this;

    for (int i = 0; i < entries.length; i++) {
      String c = entries[i];
      if (i == 0) {
        trueOrThrow(c.toLowerCase() == 'm',
            FormatException('Path must start with "m" or "M"', path));
        continue;
      }

      var hardened = (c.length > 1) && (c[c.length - 1] == "'");
      var childIndex = int.parse(c.replaceAll("'", ''),
          radix: 10); // & (HARDENED_OFFSET - 1)
      trueOrThrow(
          childIndex < HARDENED_OFFSET, FormatException('Invalid index', path));
      if (hardened) childIndex += HARDENED_OFFSET;

      hdKey = hdKey.deriveChild(childIndex);
    }

    return hdKey;
  }

  HDKey deriveChild(int index) {
    bool isHardened = index >= HARDENED_OFFSET;
    ByteData indexByteData = ByteData(4);
    indexByteData.setUint32(0, index);

    ByteData data;

    if (isHardened) {
      // Hardened child
      trueOrThrow(
          this.privateKey != null,
          FormatException(
              'Could not derive hardened child key', this.privateKey));

      ByteData zb = ByteData(1);
      List<int> tmpPk = Uint8List.view(zb.buffer).toList();
      tmpPk.addAll(this.privateKey!.toList());

      List<int> tmpData = [];
      tmpData.addAll(tmpPk); // or we can use tmpData.addAll(pk);
      tmpData.addAll(Uint8List.view(indexByteData.buffer).toList());
      data = ByteData.view(Uint8List.fromList(tmpData).buffer);
    } else {
      // Normal child
      // data = serP(point(kpar)) || ser32(index)
      //      = serP(Kpar) || ser32(index)
      List<int> pk = this.publicKey!.toList(growable: true);
      pk.addAll(Uint8List.view(indexByteData.buffer));
      data = ByteData.view(Uint8List.fromList(pk).buffer);
    }

    Uint8List I = sha512(Uint8List.view(data.buffer), this.chainCode);
    Uint8List IL = I.sublist(0, 32);
    Uint8List IR = I.sublist(32);

    HDKey hd = new HDKey(this.versions);

    // Private parent key -> private child key
    if (this.privateKey != null) {
      // ki = parse256(IL) + kpar (mod n)
      try {
        hd.privateKey = privateKeyTweakAdd(this.privateKey!, IL);
        // throw if IL >= n || (privateKey + IL) === 0
      } catch (err) {
        print("privateKeyTweakAdd error: ${err.toString()}");
        // In case parse256(IL) >= n or ki == 0, one should proceed with the next value for i
        return this.deriveChild(index + 1);
      }
      // Public parent key -> public child key
    } else {
      // Ki = point(parse256(IL)) + Kpar
      //    = G*IL + Kpar
      try {
        hd.publicKey = publicKeyTweakAdd(this.publicKey!, IL, true);
        // throw if IL >= n || (g**IL + publicKey) is infinity
      } catch (err) {
        print("publicKeyTweakAdd error: ${err.toString()}");
        // In case parse256(IL) >= n or Ki is the point at infinity, one should proceed with the next value for i
        return this.deriveChild(index + 1);
      }
    }

    hd.chainCode = IR;
    hd.depth = this.depth + 1;
    hd.parentFingerprint = this.fingerprint;
    hd.index = index;

    return hd;
  }

  //endregion

  //region Crypto Methods
  Uint8List sign(Uint8List hash) {
    trueOrThrow(this.privateKey != null,
        FormatException('Private key is null.', this.privateKey));
    return eccSign(this.privateKey!, hash);
  }

  bool verify(Uint8List hash, Uint8List signature) {
    if (this.publicKey == null) throw ('Public key not found.');
    return eccSignedHashVerify(this.publicKey!, hash, signature);
  }

  //endregion

  HDKey wipePrivateData() {
    if (this._privateKey != null) {
      this._privateKey?.setAll(0, Uint8List(this._privateKey!.length));
    }
    this._privateKey = null;
    return this;
  }

  Uint8List _serialize(int version, Uint8List key) {
    // => version(4) || depth(1) || fingerprint(4) || index(4) || chain(32) || key(33)
    ByteData byteData = ByteData(LEN);
    byteData.setUint32(0, version);
    byteData.setUint8(4, this.depth);

    int fingerprint = this.depth > 0 ? this.parentFingerprint : 0x00000000;
    byteData.setUint32(5, fingerprint);
    byteData.setUint32(9, this.index);

    List<int> buffer = [];
    buffer.addAll(Uint8List.view(byteData.buffer).toList());
    buffer.setAll(13, this.chainCode!.toList());
    buffer.setAll(45, key.toList());

    return Uint8List.fromList(buffer);
  }

  @override
  bool operator ==(covariant HDKey other) {
    return this.extendedKeys == other.extendedKeys;
  }
}
